stage2.1_traited_script/traited_face_detect.py

```diff
@@ -0,0 +1,90 @@
+""" Script analyzing an image, detecting human faces inside it, and printing
+EXIF data about it.
+"""
+import PIL.Image
+from PIL.ExifTags import TAGS
+from skimage import data
+from skimage.feature import Cascade
+import matplotlib.pyplot as plt
+from matplotlib import patches
+from os.path import join
+import numpy as np
+
+# ETS imports
+from traits.api import Dict, File, HasStrictTraits, List, observe
+
+
+class ImageFile(HasStrictTraits):
+    """ Model to hold an image file.
+    """
+    filepath = File
+
+    faces = List
+
+    metadata = Dict
+
+    def to_array(self):
+        with PIL.Image.open(self.filepath) as img:
+            return np.asarray(img)
+
+    @observe("filepath")
+    def update_metadata(self, event):
+
+        with PIL.Image.open(self.filepath) as img:
+            exif = img._getexif()
+        self.metadata = {TAGS[k]: v for k, v in exif.items()
+                         if k in TAGS}
+
+    def detect_faces(self):
+        # Load the trained file from the module root.
+        trained_file = data.lbp_frontal_face_cascade_filename()
+
+        # Initialize the detector cascade.
+        detector = Cascade(trained_file)
+
+        detected = detector.detect_multi_scale(img=self.to_array(),
+                                               scale_factor=1.2,
+                                               step_ratio=1,
+                                               min_size=(60, 60),
+                                               max_size=(600, 600))
+        self.faces = detected
+
+        self.metadata["Number of faces detected"] = len(detected)
+
+
+# Select image file -----------------------------------------------------------
+
+image_path = join("..", "sample_images", "IMG-0311_xmas_2020.JPG")
```

Comment:
I think we should keep a bare minimum of code in the main namespace. This should be a function.

Reply:
Not sure I follow. A function with what in it? I was trying to keep the script almost the same, weaving in the Trait class to replace its model parts and leaving the rest as-is.

Reply:
What I mean is image_path and the like should be inside a function or inside an `if __name__ == '__main__':` block.